apps/prairielearn/src/components/SyncErrorsAndWarnings.html.tsx

```diff
@@ -0,0 +1,184 @@
+import { ansiToHtml } from '../lib/chalk.js';
+import type { StaffCourse, StaffCourseInstance } from '../lib/client/safe-db-types.js';
+import { type Assessment, type Question } from '../lib/db-types.js';
+
+export function CourseSyncErrorsAndWarnings({
+  authz_data,
+  course,
+  urlPrefix,
+}: {
+  authz_data: { has_course_instance_permission_edit: boolean };
+  course: StaffCourse;
+  urlPrefix: string;
+}) {
+  return (
+    <SyncErrorsAndWarnings
+      authz_data={authz_data}
+      exampleCourse={course.example_course}
+      syncErrors={course.sync_errors}
+      syncWarnings={course.sync_warnings}
+      fileEditUrl={`${urlPrefix}/course_admin/file_edit/infoCourse.json`}
+      context={'course'}
+    />
+  );
+}
+
+export function QuestionSyncErrorsAndWarnings({
+  authz_data,
+  question,
+  course,
+  urlPrefix,
+}: {
+  authz_data: { has_course_instance_permission_edit: boolean };
+  question: Question;
+  course: StaffCourse;
+  urlPrefix: string;
+}) {
+  return (
+    <SyncErrorsAndWarnings
+      authz_data={authz_data}
+      exampleCourse={course.example_course}
+      syncErrors={question.sync_errors}
+      syncWarnings={question.sync_warnings}
+      fileEditUrl={`${urlPrefix}/question/${question.id}/file_edit/questions/${question.qid}/info.json`}
+      context={'question'}
+    />
+  );
+}
+
+export function CourseInstanceSyncErrorsAndWarnings({
+  authz_data,
+  courseInstance,
+  course,
+  urlPrefix,
+}: {
+  authz_data: { has_course_instance_permission_edit: boolean };
+  courseInstance: StaffCourseInstance;
+  course: StaffCourse;
+  urlPrefix: string;
+}) {
+  return (
+    <SyncErrorsAndWarnings
+      authz_data={authz_data}
+      exampleCourse={course.example_course}
+      syncErrors={courseInstance.sync_errors}
+      syncWarnings={courseInstance.sync_warnings}
+      fileEditUrl={`${urlPrefix}/instance_admin/file_edit/courseInstances/${courseInstance.short_name}/infoCourseInstance.json`}
+      context={'course instance'}
+    />
+  );
+}
+
+export function AssessmentSyncErrorsAndWarnings({
+  authz_data,
+  assessment,
+  courseInstance,
+  course,
+  urlPrefix,
+}: {
+  authz_data: { has_course_instance_permission_edit: boolean };
+  assessment: Assessment;
+  courseInstance: StaffCourseInstance;
+  course: StaffCourse;
+  urlPrefix: string;
+}) {
+  return (
+    <SyncErrorsAndWarnings
+      authz_data={authz_data}
+      exampleCourse={course.example_course}
+      syncErrors={assessment.sync_errors}
+      syncWarnings={assessment.sync_warnings}
+      fileEditUrl={`${urlPrefix}/assessment/${assessment.id}/file_edit/courseInstances/${courseInstance.short_name}/assessments/${assessment.tid}/infoAssessment.json`}
+      context={'assessment'}
+    />
+  );
+}
+
+function SyncErrorsAndWarnings({
+  authz_data,
+  exampleCourse,
+  syncErrors,
+  syncWarnings,
+  fileEditUrl,
+  context,
+}: {
+  authz_data: { has_course_instance_permission_edit: boolean };
```

Comment:
For lack of a better place to put this comment, this PR will need handling of `authz_data` to ensure that we don't leak private information to the client.

That said: maybe this component should always be server-rendered? Certainly I don't think there's a need for this to ever render on the client, as it won't be interactive. In this PR that's already the case, though I see you're trying to render it on the client in #12211.

Reply:
I would be happen to update all pages to have resLocals flow through some sort of zod validation.

Reply:
In the interest of moving along this PR and the one that depends on in, I think it's easiest to just keep this on the server for now.